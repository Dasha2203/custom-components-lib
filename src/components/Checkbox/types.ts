import { InputHTMLAttributes } from 'react';

export type CheckboxProps = Omit<
  InputHTMLAttributes<HTMLInputElement>,
  'type'
> & {
  label?: string;
};
