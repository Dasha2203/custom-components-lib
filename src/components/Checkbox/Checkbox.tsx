import React from 'react';
import clsx from 'clsx';
import * as classes from './style.module.scss';
import { CheckboxProps } from './types';
import CheckedIcon from '@/icons/CheckedIcon';
import UnCheckedIcon from '@/icons/UnCheckedIcon';

const Checkbox = ({
  label,
  checked,
  disabled,
  onChange,
  className,
  style,
  ...props
}: CheckboxProps) => {
  return (
    <label
      className={clsx(
        classes.checkbox,
        checked && classes.checkbox_checked,
        disabled && classes.checkbox_disabled,
        className
      )}
      style={style}
    >
      <span className={classes.checkbox__wrap}>
        <input
          type="checkbox"
          className={classes.checkbox__input}
          checked={checked}
          onChange={onChange}
          disabled={disabled}
          aria-checked={checked}
          {...props}
        />

        {checked ? (
          <CheckedIcon
            className={classes.checkbox__icon}
            focusable="false"
            aria-hidden="true"
          />
        ) : (
          <UnCheckedIcon
            className={classes.checkbox__icon}
            focusable="false"
            aria-hidden="true"
          />
        )}
      </span>
      {label && <span className={classes.checkbox__label}>{label}</span>}
    </label>
  );
};

export default Checkbox;
