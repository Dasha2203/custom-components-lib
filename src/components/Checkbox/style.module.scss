.checkbox {
  position: relative;
  padding: 9px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  cursor: pointer;
  user-select: none;
  color: rgba(0, 0, 0, 0.6);

  &:hover:not(.checkbox_disabled) {
    .checkbox__wrap {
      color: rgba(0, 0, 0, 0.4);
    }
  }

  &_checked {
    .checkbox__wrap {
      color: rgb(25, 118, 210);
    }

    &:hover:not(.checkbox_disabled) {
      .checkbox__wrap {
        color: rgb(21, 87, 153);
      }
    }
  }

  &_disabled {
    cursor: default;
    .checkbox__wrap {
      color: rgba(0, 0, 0, 0.26);
    }
    .checkbox__label {
      color: rgba(0, 0, 0, 0.38);
    }
  }

  &__wrap {
    position: relative;
    padding: 9px;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    -webkit-tap-highlight-color: transparent;
    background-color: transparent;
    border-radius: 50%;
    color: rgba(0, 0, 0, 0.6);
    transition: color 150ms cubic-bezier(0.4, 0, 0.2, 1);
  }

  &__input {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;
    opacity: 0;
    width: 100%;
    height: 100%;
    cursor: inherit;
  }

  &__icon {
    display: inline-block;
    width: 1em;
    height: 1em;
    flex-shrink: 0;
    user-select: none;
    transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    fill: currentColor;
    font-size: 1.5rem;
  }

  &__label {
    font-weight: 400;
    font-size: 1rem;
    line-height: 1.5;
    letter-spacing: 0.00938em;
  }
}
