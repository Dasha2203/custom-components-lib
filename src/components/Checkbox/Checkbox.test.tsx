import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import Checkbox from './Checkbox';

describe('Testing Checkbox component', () => {
  test('Should render component', () => {
    render(<Checkbox defaultChecked />);

    const checkbox = screen.getByRole('checkbox');

    expect(checkbox).toBeInTheDocument();
  });

  test('Should be checked', () => {
    render(<Checkbox defaultChecked />);

    const checkbox = screen.getByRole('checkbox');

    expect(checkbox).toBeChecked();
  });

  test('Should be unchecked', () => {
    render(<Checkbox defaultChecked={false} />);

    const checkbox = screen.getByRole('checkbox');

    expect(checkbox).not.toBeChecked();
  });

  test('Should render label', () => {
    const label = 'Label';
    render(<Checkbox defaultChecked={false} label={label} />);

    const checkbox = screen.getByRole('checkbox');
    const labelEl = screen.getByText(label);

    expect(checkbox).not.toBeChecked();
    expect(labelEl).toBeInTheDocument();
  });

  test('should be disabled when disabled prop is true', () => {
    render(<Checkbox disabled defaultChecked />);

    const checkbox = screen.getByRole('checkbox');
    expect(checkbox).toBeDisabled();
  });

  test('Checkbox toggles checked state on click', () => {
    const handleChange = jest.fn();
    let checkedState = false;
    const checkedIconId = 'checked-icon';
    const ariaCheckedAttr = 'aria-checked';

    const { rerender } = render(
      <Checkbox
        checked={checkedState}
        onChange={() => {
          checkedState = !checkedState;
          handleChange();
        }}
      />
    );

    const checkbox = screen.getByRole('checkbox');

    expect(checkbox.getAttribute(ariaCheckedAttr)).toBe('false');
    expect(screen.queryByTestId(checkedIconId)).toBeNull();

    fireEvent.click(checkbox);

    expect(handleChange).toHaveBeenCalledTimes(1);

    rerender(
      <Checkbox checked={checkedState} onChange={() => handleChange()} />
    );

    expect(checkbox.getAttribute(ariaCheckedAttr)).toBe('true');
    expect(screen.queryByTestId(checkedIconId)).not.toBeNull();
  });
});

describe('Snapshots of Checkbox', () => {
  test('Checked Checkbox', () => {
    const { asFragment } = render(<Checkbox defaultChecked={true} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Unchecked Checkbox', () => {
    const { asFragment } = render(<Checkbox defaultChecked={false} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Should render correctly a disabled and checked checkbox', () => {
    const { asFragment } = render(<Checkbox checked={true} disabled />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Should render correctly a disabled and unchecked checkbox', () => {
    const { asFragment } = render(<Checkbox checked={false} disabled />);

    expect(asFragment()).toMatchSnapshot();
  });
});
