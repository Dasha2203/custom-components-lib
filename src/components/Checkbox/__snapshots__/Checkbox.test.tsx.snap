// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots of Checkbox Checked Checkbox 1`] = `
<DocumentFragment>
  <label
    class="checkbox"
    tabindex="0"
  >
    <span
      class="checkbox__wrap"
    >
      <input
        checked=""
        class="checkbox__input"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="checkbox__icon"
        data-testid="unchecked-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </span>
  </label>
</DocumentFragment>
`;

exports[`Snapshots of Checkbox Should render correctly a disabled and checked checkbox 1`] = `
<DocumentFragment>
  <label
    class="checkbox checkbox_checked checkbox_disabled"
    tabindex="0"
  >
    <span
      class="checkbox__wrap"
    >
      <input
        aria-checked="true"
        checked=""
        class="checkbox__input"
        disabled=""
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="checkbox__icon"
        data-testid="checked-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
        />
      </svg>
    </span>
  </label>
</DocumentFragment>
`;

exports[`Snapshots of Checkbox Should render correctly a disabled and unchecked checkbox 1`] = `
<DocumentFragment>
  <label
    class="checkbox checkbox_disabled"
    tabindex="0"
  >
    <span
      class="checkbox__wrap"
    >
      <input
        aria-checked="false"
        class="checkbox__input"
        disabled=""
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="checkbox__icon"
        data-testid="unchecked-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </span>
  </label>
</DocumentFragment>
`;

exports[`Snapshots of Checkbox Unchecked Checkbox 1`] = `
<DocumentFragment>
  <label
    class="checkbox"
    tabindex="0"
  >
    <span
      class="checkbox__wrap"
    >
      <input
        class="checkbox__input"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="checkbox__icon"
        data-testid="unchecked-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </span>
  </label>
</DocumentFragment>
`;
